board_and_pieces.py

```diff
@@ -101,4 +202,15 @@ def __str__(self) -> None:
                     arr.append("-")
             s += " ".join(arr) + "\n"
         return s
- 
\ No newline at end of file
+
+    def board_copy(self):
```

Comment:
Insteads of piece_copy(), and board_copy(), why don't you implement the __deepcopy__  special method for all pieces and the Board? That way, you can just do

`
b = Board()
b_copy = b.deepcopy()
`

see https://docs.python.org/3/library/copy.html
> In order for a class to define its own copy implementation, it can define special methods __copy__() and __deepcopy__().